### PR TITLE
Enh/shoot state deletion

### LINKED ISSUE #3962: Shoot cannot be created due to a a race condition with the shoot state
**How to categorize this issue?**

/kind bug
/priority 3

**What happened**:

A cluster that was running several days was deleted.
**Directly** after the cluster was deleted a new cluster with the same name has been created. 

Gardener was unable to create the new shoot due to a already existing shoot state resource with a different owner. (see below for a concrete error message).

_my assumption_:
I suspect that gardener relies on the k8s garbage collection for the shoot state deletion after the owning shoot is deleted.
But if the garbage collection does not happen before a shoot with the same name is created, the old state conflicts with the new shoot.

Detailed err message:
```
Flow "Shoot cluster deletion" encountered task errors: [task "Ensuring that ShootState exists" failed: retry failed with context deadline exceeded, last error: ShootState.core.gardener.cloud "myshoot" is invalid: metadata.ownerReferences: Invalid value: []v1.OwnerReference{v1.OwnerReference{APIVersion:"core.gardener.cloud/v1beta1", Kind:"Shoot", Name:"myshoot", UID:"7ca9d19a-4d5c-43ef-8b9e-fa80dafc179a", Controller:(*bool)(0xc0a3bc3a80), BlockOwnerDeletion:(*bool)(0xc0a3bc3a81)}, v1.OwnerReference{APIVersion:"core.gardener.cloud/v1beta1", Kind:"Shoot", Name:"myshoot", UID:"b75f622a-469c-4a23-9747-d054d3715a11", Controller:(*bool)(0xc0a3bc3a82), BlockOwnerDeletion:(*bool)(0xc0a3bc3a83)}}: Only one reference can have Controller set to true. Found "true" in references for myshoot and myshoot] Operation will be retried.
```

**What you expected to happen**:

The shoot state should get cleaned up before the shoot resource is deleted.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a shoot named `shoot-abc`
2. Delete that shoot and directly afterwards create a new shoot with the same name.

**Anything else we need to know?**:

**Environment**:

- Gardener version: v1.21.0
- Kubernetes version (use `kubectl version`): 1.20.5
- Cloud provider or hardware configuration: GCP

### PR DESCRIPTION
**Pull request by BeckerMax**
**How to categorize this PR?**

/area control-plane

**What this PR does / why we need it**:

Before this change the Shootstate was deleted via owner references.
This led to problems when deleting and creating a Shoot with the same name
in quick succession as our `ShootStateDeletionValidator` admission plugin disallowed the deletion of the
ShootState if a Shoot with the same name exists.

Now the admission plugin is removed and the ShootState is deleted
explicitly as part of the shoot deletion flow. To protect from unintentional
deletion a deletion confirmation is needed to delete the ShootState.

**Which issue(s) this PR fixes**:
Fixes #3962

**Special notes for your reviewer**:

I would like to discuss https://github.com/gardener/gardener/compare/master...BeckerMax:enh/shoot-state-deletion?expand=1#diff-1b66a0fabd3ab9fcedbaaa435a65b221b2208c2874ce24f07692152e2e32a574R521-R526

I understood from the code that implicitly with every run of `EnsureShootStateExists` the `o.ShootState` and `o.Shoot.ETCDEncryption` got updated. I tried to mimic the behaviour with the new implementation. But would like a second opinion here. Also I find it unintuitive that ensureShootStateExists updates o.Shoot.ETCDEncryption. Maybe a different step or method name would be better?
WDYT?

Also I needed to use the a __versioned clientSet__ to interact with the ShootState as the __internal ClienSet__ will default to trying to prioritize using  the v1beta1 version. I also considered to change the priority order so that v1alpha1 is used by the internal ClientSet for the ShootState but did not find a way to do so.

**Release note**:

```breaking operator
The admission plugin `ShootStateDeletionValidator` is removed. Explicitly enabling or disabling it via the gardener-api-server will cause the gardener-api-server to fail to start. This fixes an error caused by a not-in-time cleaned up ShootState resulting in Shoot creation to fail if a Shoot was deleted and created with the same name in quick succession.
```

/squash

### CODE REVIEW

**Comment by timebertt** [comment]
> I would like to discuss https://github.com/gardener/gardener/compare/master...BeckerMax:enh/shoot-state-deletion?expand=1#diff-1b66a0fabd3ab9fcedbaaa435a65b221b2208c2874ce24f07692152e2e32a574R521-R526
> 
> I understood from the code that implicitly with every run of `EnsureShootStateExists` the `o.ShootState` and `o.Shoot.ETCDEncryption` got updated. I tried to mimic the behaviour with the new implementation. But would like a second opinion here. Also I find it unintuitive that ensureShootStateExists updates o.Shoot.ETCDEncryption. Maybe a different step or method name would be better?
> WDYT?

I don't see a big problem with this. The step reads the ShootState and extracts some information that is relevant somewhere else. 
You can of course also extract the `ETCDEncryption` information somewhere else instead of in `EnsureShootStateExists`, I don't mind. But the `ETCDEncryption` handling seems to be a bit weird/unclear to me, so I guess you probably don't want to touch it in this PR :)

File: plugin/pkg/global/deletionconfirmation/admission.go [hunk 1]
```diff
@@ -32,6 +35,7 @@ import (
 	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
 	admissioninitializer "github.com/gardener/gardener/pkg/apiserver/admission/initializer"
 	"github.com/gardener/gardener/pkg/client/core/clientset/internalversion"
+	"github.com/gardener/gardener/pkg/client/core/clientset/versioned"
```
Comment (timebertt) [inline plugin/pkg/global/deletionconfirmation/admission.go]:
```suggestion
	 externalcoreclientset "github.com/gardener/gardener/pkg/client/core/clientset/versioned"
```
for consistency

File: plugin/pkg/global/deletionconfirmation/admission.go [hunk 2]
```diff
@@ -93,11 +101,24 @@ func (d *DeletionConfirmation) SetInternalCoreInformerFactory(f coreinformers.Sh
...
+
+	readyFuncs = append(readyFuncs, shootStateInformer.Informer().HasSynced)
+}
+
 // SetInternalCoreClientset gets the clientset from the Kubernetes client.
 func (d *DeletionConfirmation) SetInternalCoreClientset(c internalversion.Interface) {
 	d.gardenCoreClient = c
 }
 
+// SetInternalCoreClientset gets the clientset from the Kubernetes client.
```
Comment (timebertt) [inline plugin/pkg/global/deletionconfirmation/admission.go]:
```suggestion
// SetExternalCoreClientset gets the clientset from the Kubernetes client.
```

File: plugin/pkg/global/deletionconfirmation/admission.go [hunk 3]
```diff
@@ -168,6 +197,26 @@ func (d *DeletionConfirmation) Validate(ctx context.Context, a admission.Attribu
 		}
 		checkFunc = gutil.CheckIfDeletionIsConfirmed
 
+	case v1alpha1.Kind("ShootState"):
+		listFunc = func() ([]client.Object, error) {
+			list, err := d.shootStateLister.List(labels.Everything())
```
Comment (timebertt) [inline plugin/pkg/global/deletionconfirmation/admission.go:202]: I just realized, having a watch for all `ShootState`s is probably not a good idea, as it will increase the api server's memory footprint even more (`ShootState`s are large resources).
We should probably use a metadata-only watch/client here, which is sufficient for this admission plugin.

File: pkg/gardenlet/controller/shoot/shoot_control_delete.go [hunk 4]
```diff
@@ -556,11 +556,16 @@ func (c *Controller) runDeleteShootFlow(ctx context.Context, o *operation.Operat
...
-		_ = g.Add(flow.Task{
+		waitUntilNamespaceDeleted = g.Add(flow.Task{
 			Name:         "Waiting until shoot namespace in Seed has been deleted",
 			Fn:           botanist.WaitUntilSeedNamespaceDeleted,
 			Dependencies: flow.NewTaskIDs(deleteNamespace),
 		})
+		_ = g.Add(flow.Task{
+			Name:         "Deleting Shoot State",
+			Fn:           botanist.DeleteShootState,
+			Dependencies: flow.NewTaskIDs(syncPoint, waitUntilNamespaceDeleted),
```
Comment (timebertt) [inline pkg/gardenlet/controller/shoot/shoot_control_delete.go]:
```suggestion
			 Dependencies: flow.NewTaskIDs(waitUntilNamespaceDeleted),
```
Dependency to `syncPoint` is implicitly given already.

As said in person, I don't have a strong opinion on when to do add this step in the flow, but I think, doing it in parallel to `WaitUntilSeedNamespaceDeleted` is also fine.

File: pkg/operation/operation.go [hunk 5]
```diff
@@ -502,30 +500,49 @@ func (o *Operation) InjectShootShootImages(values map[string]interface{}, names
...
+		err        error
+		shootState = &gardencorev1alpha1.ShootState{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      o.Shoot.Info.Name,
+				Namespace: o.Shoot.Info.Namespace,
+			},
+		}
+	)
+
+	if err = o.K8sGardenClient.Client().Create(ctx, shootState); err != nil {
```
Comment (timebertt) [inline pkg/operation/operation.go:515]: Probably switch the order of `Get` and `Create`, as `Get` happens a lot more frequently?

**Review by timebertt — CHANGES_REQUESTED** [review]
Generally, looks good.
Though, during the review scalability concerns arose...

File: plugin/pkg/global/deletionconfirmation/admission.go (hunk 3 above)
Comment (rfranzke) [reply plugin/pkg/global/deletionconfirmation/admission.go:202]: @timebertt I assume the API server already watches `ShootState`s anyways, doesn't it? How does it respond to clients reading `ShootState`s?

File: docs/concepts/apiserver_admission_plugins.md [hunk 6]
```diff
@@ -109,14 +109,6 @@ by setting `spec.kubernetes.verticalPodAutoscaler.enabled=true` for newly create
 Already existing Shoots and new Shoots that explicitly disable VPA (`spec.kubernetes.verticalPodAutoscaler.enabled=false`)
 will not be affected by this admission plugin.
 
-## `ShootStateDeletionValidator`
```
Comment (rfranzke) [inline docs/concepts/apiserver_admission_plugins.md:112]: Please adapt the `## DeletionConfirmation` section

File: pkg/operation/operation.go (hunk 5 above)
Comment (rfranzke) [reply pkg/operation/operation.go:515]: This won't work as the seed authorizer won't allow a `Get` if the object does not exist yet.
Let's
/hold
this PR until #4044 is merged where this function is adapted, see https://github.com/gardener/gardener/pull/4044/files#diff-1b66a0fabd3ab9fcedbaaa435a65b221b2208c2874ce24f07692152e2e32a574L530.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/operation/operation.go (hunk 5 above)
Comment (timebertt) [reply pkg/operation/operation.go:515]: Ah right. Forgot about that again, sorry 🙈

File: plugin/pkg/global/deletionconfirmation/admission.go (hunk 3 above)
Comment (timebertt) [reply plugin/pkg/global/deletionconfirmation/admission.go:202]: Hmm, yes indeed. In the old `ShootState` plugin, g-api is already watching `ShootState`s: https://github.com/gardener/gardener/blob/2fb3b07c5c741b432889ce155a89a82cd88532b1/plugin/pkg/shootstate/validator/admission.go#L81

Still, this would be a very good opportunity for improvement.

> How does it respond to clients reading `ShootState`s?

The API server doesn't watch (via the k8s API) all resources by default. It only opens watches to etcd, if some client opens an API watch. But as long as clients only read from the API, there is no watch connection, IIRC.

**Comment by rfranzke** [comment]
OK, #4044 is merged, so
/unhold
/rebase

File: pkg/gardenlet/controller/shoot/shoot_control_delete.go (hunk 4 above)
Comment (plkokanov) [reply pkg/gardenlet/controller/shoot/shoot_control_delete.go]: Yep, think it would work if you do it in parallel to `WaitUntilSeedNamespaceDeleted` and just add a dependency to `deleteNamespace`

File: plugin/pkg/global/deletionconfirmation/admission.go (hunk 3 above)
Comment (BeckerMax) [reply plugin/pkg/global/deletionconfirmation/admission.go:202]: @timebertt and me agreed upon implementing a watch on the metadata of the ShootState. Due to their size ShootStates watches impose a signifacant amount of memory usage on APIservers in landscapes with a lot of Shoots.

Comment (timebertt) [reply plugin/pkg/global/deletionconfirmation/admission.go:202]: @BeckerMax and I concluded the following: although we see a significant benefit in using metadata-only watches or not watching shootstates at all, we don't want to do it in this PR, as shootstates were already watched before, so this PR doesn't change anything wrt memory consumption.
We tried the following approaches:
- using the c-r cache for starting metadata-only watches: problem here is, that we can't use the loopback rest config in g-api, but with the kube-apiserver rest config we have a cyclic dependency on startup: g-api is not discoverable until the startup has finished, but startup only finishes once the `core.gardener.cloud` API group is discoverable in order to construct a metada-only watch for shootstates
- completely leave out cache lookups and only rely on live lookups: this means we would need to list all shootstates for `DELETECOLLECTION` calls, which will again be a huge performance drawback

The last option was to use `client-go` style metadata-only watches/clients, but we don't have any reference usage for it in g/g yet, so we didn't try it out, but it's probably the way to go here.

Long story short, we should definitely improve the memory footprint of g-api by not watching shootstates directly, but this is not quite straight-forward so we want to do it in another PR.

Comment (timebertt) [reply plugin/pkg/global/deletionconfirmation/admission.go:202]: Opened https://github.com/gardener/gardener/issues/4106 regarding this improvement. Resolving this.

**Comment by BeckerMax** [comment]
I incorporated your feedback. I accidentally merged the feedback it into the other changes (no seperate feedback commit) while rebasing. PTAL

File: pkg/apiserver/plugins.go [hunk 7]
```diff
@@ -59,7 +57,6 @@ var (
 		deletionconfirmation.PluginName,            // DeletionConfirmation
 		openidconnectpreset.PluginName,             // OpenIDConnectPreset
 		clusteropenidconnectpreset.PluginName,      // ClusterOpenIDConnectPreset
-		shootstatedeletionvalidator.PluginName,     // ShootStateDeletionValidator
```
Comment (timebertt) [inline pkg/apiserver/plugins.go:62]: Can you mention the removal of the `ShootStateDeletionValidator` admission plugin in the release note?
If someone enables/disables the plugin explicitly, g-api will refuse to start.

File: pkg/operation/operation.go (hunk 5 above)
Comment (timebertt) [inline pkg/operation/operation.go:515]: Nit: the idiomatic way in go to deal with such errors, that are only scoped to the `if` statement, would be via the short-hand var declaration (so no need for the var declaration above):
```suggestion
	if err := o.K8sGardenClient.Client().Create(ctx, shootState); err != nil {
```

**Review by timebertt — DISMISSED** [review]
Nice, thank you!
/lgtm

**Review by rfranzke — DISMISSED** [review]
/lgtm

File: pkg/operation/operation.go (hunk 5 above)
Comment (BeckerMax) [reply pkg/operation/operation.go:515]: I did it because the error declaration is used below:
https://github.com/gardener/gardener/pull/4100/files/805237f0aff31aaf96475d5d3a26bae812f430b5#diff-1b66a0fabd3ab9fcedbaaa435a65b221b2208c2874ce24f07692152e2e32a574R526

Comment (BeckerMax) [reply pkg/operation/operation.go:515]: I generally agree with you. A more narrow scoped errors is better.

I need the error variable declaration anyways for https://github.com/gardener/gardener/pull/4100/files/805237f0aff31aaf96475d5d3a26bae812f430b5#diff-1b66a0fabd3ab9fcedbaaa435a65b221b2208c2874ce24f07692152e2e32a574R526

That's why I would leave it like that there.

**Comment by BeckerMax** [comment]
After a discussion with @timebertt we decided to *not* deliver this as a hotfix as it includes breaking changes. See the release note.

File: pkg/apiserver/plugins.go (hunk 7 above)
Comment (BeckerMax) [reply pkg/apiserver/plugins.go:62]: done

**Review by rfranzke — APPROVED** [review]

**Review by timebertt — APPROVED** [review]
